code/model/model.h

```diff
@@ -951,37 +951,53 @@ extern void submodel_rotate(bsp_info *sm, submodel_instance *smi);
+// Given a point in a submodel's local frame of reference, transform it to a global frame of reference, taking into account submodel rotations.
+// If objorient and objpos are supplied, this will be world space; otherwise it will be the model's space.
+extern void model_instance_local_to_global_point(vec3d *outpnt, const vec3d *mpnt, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const matrix *objorient = nullptr, const vec3d *objpos = nullptr);
+
+// Given a direction (or normal) in a submodel's local frame of reference, transform it to a global frame of reference.
+// If objorient and objpos are supplied, this will be world space; otherwise it will be the model's space.
+extern void model_local_to_global_dir(vec3d *out_dir, const vec3d *in_dir, int model_num, int submodel_num, const matrix *objorient = nullptr);
+// Given a direction (or normal) in a submodel's local frame of reference, transform it to a global frame of reference.
+// If objorient and objpos are supplied, this will be world space; otherwise it will be the model's space.
+extern void model_local_to_global_dir(vec3d *out_dir, const vec3d *in_dir, const polymodel *pm, int submodel_num, const matrix *objorient = nullptr);
+
+// Given a direction (or normal) in a submodel's local frame of reference, transform it to a global frame of reference, taking into account submodel rotations..
+// If objorient and objpos are supplied, this will be world space; otherwise it will be the model's space.
+extern void model_instance_local_to_global_dir(vec3d *out_dir, const vec3d *in_dir, int model_instance_num, int submodel_num, const matrix *objorient = nullptr, bool use_submodel_parent = false);
+// Given a direction (or normal) in a submodel's local frame of reference, transform it to a global frame of reference, taking into account submodel rotations..
+// If objorient and objpos are supplied, this will be world space; otherwise it will be the model's space.
+extern void model_instance_local_to_global_dir(vec3d *out_dir, const vec3d *in_dir, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const matrix *objorient = nullptr);
+
+// Given a point in world coordinates, transform it to a submodel's local frame of reference.
 // This is special purpose code, specific for model collision.
 // NOTE - this code ASSUMES submodel is 1 level down from hull (detail[0])
-void world_find_model_instance_point(vec3d *out, vec3d *world_pt, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const matrix *orient, const vec3d *pos);
+// TODO - fix this function to work for all submodel levels
+extern void model_instance_world_to_local_point(vec3d *out, vec3d *world_pt, const polymodel *pm, const polymodel_instance *pmi, int submodel_num, const matrix *orient, const vec3d *pos);
```

Comment:
Minor, but this should be "global" instead of "world" for consistency yeah?

Reply:
This one I deliberately broke consistency.  In the other functions, "global" could mean either the ship RF or the world RF, depending on optional parameters.  But this function only transforms to world RF.

Reply:
Ah, fair.